openmmtools/alchemy.py

```diff
@@ -12,413 +12,905 @@
+
+        # Modify forces as appropriate, copying other forces without modification.
+        for reference_force in reference_system.getForces():
+            # TODO switch to functools.singledispatch when drop Python2
+            reference_force_name = reference_force.__class__.__name__
+            alchemical_force_creator_name = '_alchemically_modify_{}'.format(reference_force_name)
+            try:
+                alchemical_force_creator_func = getattr(self, alchemical_force_creator_name)
+            except AttributeError:
+                alchemical_forces = [copy.deepcopy(reference_force)]
+            else:
+                alchemical_forces = alchemical_force_creator_func(reference_force, alchemical_region)
+            for alchemical_force in alchemical_forces:
+                alchemical_system.addForce(alchemical_force)
+
+        # Record timing statistics.
+        timer.stop('Create alchemically modified system')
+        timer.report_timing()
+
+        return alchemical_system
 
     @classmethod
-    def _getSystemGlobalParameters(cls, system, prefix='lambda_'):
-        """
-        Get a list of available alchemical parameters beginning with 'prefix'
+    def get_energy_components(cls, alchemical_system, alchemical_state, positions, platform=None):
+        """Compute potential energy of the alchemical system by Force.
+
+        This can be useful for debug and analysis.
 
         Parameters
         ----------
-        prefix : str, optional, default='lambda_'
-            The prefix that all Context parameters found must begin with.
+        alchemical_system : simtk.openmm.AlchemicalSystem
+            An alchemically modified system.
+        alchemical_state : AlchemicalState
+            The alchemical state to set the Context to.
+        positions : simtk.unit.Quantity of dimension (natoms, 3)
+            Coordinates to use for energy test (units of distance).
+        platform : simtk.openmm.Platform, optional
+            The OpenMM platform to use to compute the energy. If None,
+            OpenMM tries to select the fastest available.
+
+        Returns
+        -------
+        energy_components : dict str: simtk.unit.Quantity
+            A string label describing the role of the force associated to
+            its contribution to the potential energy.
 
         """
-        platform = openmm.Platform.getPlatformByName('Reference')
+        # Find and label all forces.
+        force_labels = cls._find_force_components(alchemical_system)
+        assert len(force_labels) <= 32, ("The alchemical system has more than 32 force groups; "
+                                         "can't compute individual force component energies.")
+
+        # Create deep copy of alchemical system.
+        system = copy.deepcopy(alchemical_system)
+
+        # Separate all forces into separate force groups.
+        for force_index, force in enumerate(system.getForces()):
```

Comment:
Yeah, that's a limitation, but since this is just code for testing, I'd implement only if needed.